// config/sendProgress.js
var gcm = require('node-gcm');
var event = require('config/modelEvent');   
var googleApiKey = 'AIzaSyAlOAwGDGVgt1IOEALhg7AMb7SzTiMMKqI';

//Aggregation function
function getAmount(id_event, callback) {
    Event.aggregate([
    	{ $match: {_id: id_event }},
        { $unwind: "$usersActing" }, 	// creates a doc for every array element - each action
        { $group: {
            _id: null,
            energy: { $sum: "$usersActing.amountEnergy"  }
        }}
    ], function (err, result) {
        if (err) {
            console.log(err);
            return;
        }

        if (result.length > 0) {
        	callback({'response':result[0].energy, 'res':"true"});
	}
	else {
		callback({'response':0, 'res':"true"});
	}
        return result;
    });
}

function timer() {
    var sender = new gcm.Sender(googleApiKey);
    var message = new gcm.Message();
    
    message.addData('message', getProgress());

  	console.log(message);
  	message.delay_while_idle = 1;
  	var registrationIds = [];
  
    event.findOne({}, {}, { sort: { 'date' : -1 } }, function(err, lastEvent) {
      lastEvent.find(function(err, users) {  
  		  users.forEach(function(user) {
    			registrationIds.push(user.gcmregid);
    			console.log("GCM Key " + user.gcmregid);  
    	  });
  
  	    sender.send(message, registrationIds, 4, function (err, result) {
  			  console.log(result);
  		  });
  	  });   
    });
}

function stopFunction(timerFunction) {
    clearInterval(timerFunction);
} 

module.exports = {
  sendProgress: function (timeout, frequency) {
    	  //Calling function every minute
	  var sendingPieces = window.setInterval(function(){ timer() }, 1000*frequency);
	  
	  //Stoping function after a certain amount of time
	  window.setTimeout(function(){ stopFunction(sendingPieces); }, 1000*60*timeout);
  }
};
